nemo/collections/nlp/models/question_answering/qa_bert_model.py

```diff
@@ -0,0 +1,892 @@
+            start_logits = []
+            end_logits = []
+            for u in all_unique_ids:
+                unique_ids.extend(tensor2list(u))
+            for u in all_start_logits:
+                start_logits.extend(tensor2list(u))
+            for u in all_end_logits:
+                end_logits.extend(tensor2list(u))
+
+            eval_dataset = self._test_dl.dataset if self.trainer.testing else self._validation_dl.dataset
+            exact_match, f1, all_predictions, all_nbest = self.evaluate(
+                eval_dataset.features,
+                eval_dataset.examples,
+                eval_dataset.processor,
+                unique_ids=unique_ids,
+                start_logits=start_logits,
+                end_logits=end_logits,
+                n_best_size=self._cfg.dataset.n_best_size,
+                max_answer_length=self._cfg.dataset.max_answer_length,
+                version_2_with_negative=self._cfg.dataset.version_2_with_negative,
+                null_score_diff_threshold=self._cfg.dataset.null_score_diff_threshold,
+                do_lower_case=self._cfg.dataset.do_lower_case,
+            )
+
+        logging.info(f"{prefix} exact match {exact_match}")
+        logging.info(f"{prefix} f1 {f1}")
+
+        self.log(f'{prefix}_loss', avg_loss)
+        self.log(f'{prefix}_exact_match', exact_match)
+        self.log(f'{prefix}_f1', f1)
+
+    def test_epoch_end(self, outputs):
+        return self.validation_epoch_end(outputs)
+
+    @typecheck()
+    def forward(self, input_ids, attention_mask, token_type_ids):
+        with torch.cuda.amp.autocast():
+            hidden_states = self.bert_model(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+
+            if isinstance(hidden_states, tuple):
+                hidden_states = hidden_states[0]
+
+            logits = self.classifier(hidden_states=hidden_states)
+
+        return logits
+
+    @torch.no_grad()
+    def inference(
+        self,
+        file: str,
+        batch_size: int = 1,
+        num_samples: int = -1,
+        output_nbest_file: Optional[str] = None,
+        output_prediction_file: Optional[str] = None,
+    ):
+        """
+        Get prediction for unlabeled inference data
+
+        Args:
+            file: inference data
+            batch_size: batch size to use during inference
+            num_samples: number of samples to use of inference data. Default: -1 if all data should be used.
+            output_nbest_file: optional output file for writing out nbest list
+            output_prediction_file: optional output file for writing out predictions
+            
+        Returns:
+            model predictions, model nbest list
+        """
+
+        # store predictions for all queries in a single list
+        all_predictions = []
+        all_nbest = []
+        mode = self.training
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
```

Comment:
wouldnt recommend this - instead ask the trainer.devices since 'cuda' by default will use the first device (i.e. cuda:0) but sometimes, you can specify devices=[1] which is cuda:1

Reply:
From what I'm seeing in the inference method and the examples script for QA, you can try to call self.trainer.device? The model was initialized with the trainer and hence can call the trainer. If this doesn't work, we can also leave it as it is.